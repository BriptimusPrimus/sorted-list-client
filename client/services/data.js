import config from '../../config';

const apiEndpoint = config.data_service.api.protocol + '://'
  + config.data_service.api.host + ':'
  + config.data_service.api.port + '/'
  + config.data_service.api.root + '/';

function getRequest(endpoint) {
  function makeRequest(resolve, reject) {
    const requestConf = {
      method: 'GET',
      mode: 'cors',
      headers: new Headers({
        'Content-Type': 'application/json'
      })
    }

    fetch(endpoint, requestConf)
      .then(function(response) {
        if (response.status !== 200) {
          return {
            error: 'Service Failure',
            status: response.status
          }
        }
        return response.json();
      })
      .then(function(data) {
        if(data.error) {
          reject(data);
        } else {
          resolve(data);
        }
      })
      .catch(function(reason) {
        reject({
          error: 'Service Failure'
        });
      });
  }

  return new Promise(function(resolve, reject) {
    makeRequest(resolve, reject);
  });
}

function postRequest(enpoint, data) {
  function makeRequest(resolve, reject) {
    const requestConf = {
      method: 'POST',
      mode: 'cors',
      headers: new Headers({
        'Content-Type': 'application/json'
      }),
      body: JSON.stringify(data)
    }

    fetch(endpoint, requestConf)
      .then(function(response) {
        if (response.status !== 200) {
          return {
            error: 'Service Failure',
            status: response.status
          }
        }        
        return response.json();
      })
      .then(function(data) {
        if(data.error) {
          reject(data);
        } else {
          resolve(data);
        }
      })
      .catch(function(reason) {
        reject({
          error: 'Service Failure'
        });
      });
  }

  return new Promise(function(resolve, reject) {
    makeRequest(resolve, reject);
  });

}

function getEmployees() {
  const endpoint = apiEndpoint
    + config.data_service.api.employees_data +
    '?count=' + config.data_service.defaulCount;

  return getRequest(endpoint);
}

export default {
  getEmployees
}