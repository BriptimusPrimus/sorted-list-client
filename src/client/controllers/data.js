import dataService from '../services/data';
import {sortByColumn} from '../actions';

function sortBy(dispatchFn, column) {
  const fetchOpts = {
    sortBy: store.getState().sortBy
  };  
  
  dispatchFn(sortByColumn(column));
  dataService.getEmployees(fetchOpts)
    .then(function fullfilled(data) {
      store.dispatch(receiveData(data));
    })
    .catch(function rejected(reason) {
      console.log('response error:', reason);
    })
}

export default {
  sortBy
}