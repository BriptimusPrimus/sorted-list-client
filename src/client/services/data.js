import fetch from 'isomorphic-fetch';
import querystring from 'querystring';
import config from '../../../config';

const apiEndpoint = `${config.data_service.api.protocol}://${config.data_service.api.host}:${config.data_service.api.port}/${config.data_service.api.root}/`;

const getRequest = async function getRequest(endpoint) {
  const requestConf = {
    method: 'GET',
    mode: 'cors',
    headers: new Headers({
      'Content-Type': 'application/json'
    })
  };

  const response = await fetch(endpoint, requestConf);

  if (response.status === 404) {
    return {
      notFound: true
    };
  }

  if (response.status !== 200) {
    return {
      error: 'Service Failure',
      status: response.status
    };
  }

  const data = await response.json();
  if (data.error) {
    throw new Error(data);
  }

  return data;
};

const getEmployees = function getEmployees(options = {}) {
  const sortBy = options.sortBy || {
    column: 'codeNumber'
  };

  const params = {
    count: options.count || config.data_service.defaulCount,
    sort: sortBy.column,
    order: sortBy.orderDesc ? 'DESC' : 'ASC'
  };

  const endpoint = `${
    apiEndpoint + config.data_service.api.employees_data
  }?${querystring.stringify(params)}`;

  return getRequest(endpoint);
};

const getCustomers = function getCustomers(options = {}) {
  const sortBy = options.sortBy || {
    column: 'id'
  };

  const params = {
    count: options.count || config.data_service.defaulCount,
    sort: sortBy.column,
    order: sortBy.orderDesc ? 'DESC' : 'ASC'
  };

  const endpoint = `${
    apiEndpoint + config.data_service.api.customers_data
  }?${querystring.stringify(params)}`;

  return getRequest(endpoint);
};

const getCustomerDetails = function getCustomerDetails({ customerId } = {}) {
  const endpoint = `${
    apiEndpoint + config.data_service.api.customer_details
  }/${customerId}`;

  return getRequest(endpoint);
};

const getDisplayDetails = function getDisplayDetails({ customerId } = {}) {
  const endpoint = `${
    apiEndpoint + config.data_service.api.display_details
  }?id=${customerId}`;

  return getRequest(endpoint);
};

export { getEmployees, getCustomers, getCustomerDetails, getDisplayDetails };
