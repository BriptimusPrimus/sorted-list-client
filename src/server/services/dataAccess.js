/*
 * Data Access Service Abstraction
 */

const implementationModule = require('../../../config').dal.module;

// eslint-disable-next-line import/no-dynamic-require
const dal = require(`./${implementationModule}`);

const getCustomerDetails = async function getCustomerDetails(id) {
  try {
    const data = await dal.getCustomer(id);

    // Customer not exist
    if (data === undefined) {
      return {
        success: true,
        noDataFound: true
      };
    }

    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

const getCustomers = async function getCustomers(opts) {
  const params = {
    count: Number.isNaN(Number(opts.count)) ? 20 : Number(opts.count),
    page:
      Number.isNaN(Number(opts.page)) || opts.page < 1 ? 1 : Number(opts.page),
    sortBy: opts.sortBy,
    order: opts.order || 'ASC',
    filters: opts.filters || {}
  };

  try {
    const data = await dal.getCustomers(params);
    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

const getEmployees = async function getEmployees(opts) {
  const params = {
    count: Number.isNaN(Number(opts.count)) ? 20 : Number(opts.count),
    page:
      Number.isNaN(Number(opts.page)) || opts.page < 1 ? 1 : Number(opts.page),
    sortBy: opts.sortBy,
    order: opts.order || 'ASC',
    filters: opts.filters || {}
  };

  try {
    const data = await dal.getEmployees(params);
    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

// This object implements the bridge interface:
// interface: {
//   getCustomerDetails: async function(customerId){}
//   getCustomers: async function(options){}
//   getEmployees: async function(options){}
// }
module.exports = {
  getCustomerDetails,
  getCustomers,
  getEmployees
};
