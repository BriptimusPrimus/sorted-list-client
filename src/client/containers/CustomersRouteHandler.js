import React from 'react';
import PropTypes from 'prop-types';
import { receiveData, receiveDataError } from '../actions';
import { getCustomers } from '../services/data';
import CustomersContainer from './CustomersContainer';
import withUniversalLoad from '../lib/withUniversalLoad';

const CustomersWithUniversalLoad = withUniversalLoad(CustomersContainer, {
  dataSource: getCustomers,
  dataSourceParams: {
    sortBy: {
      inState: true
    }
  },
  successAction: receiveData,
  failureAction: receiveDataError
});

const CustomersRouteHandler = ({ initialState }) => {
  return <CustomersWithUniversalLoad initialState={initialState} />;
};

// Override loadData
CustomersRouteHandler.loadData = async function preloadData(params) {
  const data = await CustomersWithUniversalLoad.loadData(params);
  return {
    data
  };
};

export default CustomersRouteHandler;

CustomersRouteHandler.propTypes = {
  initialState: PropTypes.shape({
    data: PropTypes.shape({
      customers: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number,
          firstName: PropTypes.string,
          surname: PropTypes.string,
          email: PropTypes.string,
          addressId: PropTypes.number,
          isActive: PropTypes.bool,
          createDate: PropTypes.string,
          lastUpdate: PropTypes.string
        })
      )
    }),
    sortBy: PropTypes.shape({
      column: PropTypes.string,
      orderDesc: PropTypes.bool
    })
  })
};

CustomersRouteHandler.defaultProps = {
  initialState: {
    data: {
      customers: []
    },
    sortBy: {
      column: 'id',
      orderDesc: false
    }
  }
};
