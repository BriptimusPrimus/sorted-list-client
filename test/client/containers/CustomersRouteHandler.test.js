import React from 'react';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import CustomersRouteHandler from '../../../src/client/containers/CustomersRouteHandler';
import * as mockService from '../../../src/client/services/data';
import {
  getInitialState,
  sortedBySurname,
  sortedByIsActive,
  sortByCreateDate,
  sortedByIdDesc,
  sortedByFirstName
} from '../utils/customerDataMocks';

jest.mock('react-router-dom', () => {
  return {
    useParams: () => ({}),
    useLocation: () => ({}),
    Link: ({ children }) => {
      return children;
    }
  };
});
jest.mock('../../../src/client/services/data');

describe('CustomersRouteHandler tests', () => {
  const initialState = getInitialState();

  beforeEach(() => {
    jest.clearAllMocks();
    mockService.getCustomers.mockImplementation(async () => {
      return initialState.data;
    });
  });

  afterEach(cleanup);

  test('Renders', () => {
    const { container } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Sorts by surname', async () => {
    mockService.getCustomers.mockImplementation(async () => {
      return sortedBySurname;
    });

    const { getByText, queryByText } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(queryByText('SMITH')).not.toBeNull();

    const surnameHeader = getByText('Last Name');
    fireEvent.click(surnameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('SMITH')).toBeNull();
    expect(queryByText('ABNEY')).not.toBeNull();
  });

  test('Sort column service call fails', async () => {
    mockService.getCustomers.mockImplementation(async () => {
      return Promise.reject(new Error('boom'));
    });

    const { getByText, queryByText } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(queryByText('SMITH')).not.toBeNull();

    const surnameHeader = getByText('Last Name');
    fireEvent.click(surnameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('SMITH')).not.toBeNull();
    expect(queryByText('ABNEY')).toBeNull();
  });

  test('Sorts by isActive', async () => {
    mockService.getCustomers.mockImplementation(async () => {
      return sortedByIsActive;
    });

    const { getByText, queryByText } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(queryByText('MARY')).not.toBeNull();

    const surnameHeader = getByText('Active');
    fireEvent.click(surnameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('MARY')).toBeNull();
    expect(queryByText('ERICA')).not.toBeNull();
  });

  test('Sorts by createDate', async () => {
    mockService.getCustomers.mockImplementation(async () => {
      return sortByCreateDate;
    });

    const { getByText, queryByText } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(queryByText('MARY')).not.toBeNull();

    const surnameHeader = getByText('Joined on');
    fireEvent.click(surnameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    // Same payload as default (sorted by id)
    expect(queryByText('MARY')).not.toBeNull();
  });

  test('Sorts by id', async () => {
    mockService.getCustomers.mockImplementation(async () => {
      return sortedByIdDesc;
    });

    const { getByText, queryByText } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(queryByText('MARY')).not.toBeNull();

    const surnameHeader = getByText('ID');
    fireEvent.click(surnameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('MARY')).toBeNull();
    expect(queryByText('AUSTIN')).not.toBeNull();
  });

  test('Sorts by firstName', async () => {
    mockService.getCustomers.mockImplementation(async () => {
      return sortedByFirstName;
    });

    const { getByText, queryByText } = render(
      <CustomersRouteHandler initialState={initialState} />
    );
    expect(queryByText('MARY')).not.toBeNull();

    const surnameHeader = getByText('First Name');
    fireEvent.click(surnameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    // Result is same as initial state
    expect(queryByText('MARY')).toBeNull();
    expect(queryByText('AARON')).not.toBeNull();
  });

  test('Static load data', async () => {
    const result = await CustomersRouteHandler.loadData();
    expect(result).toEqual({
      data: initialState.data
    });
  });

  test('Renders empty data by default', () => {
    const { container, queryByText } = render(<CustomersRouteHandler />);
    expect(container.firstChild).not.toBeNull();
    expect(queryByText('MARY')).toBeNull();
  });
});
