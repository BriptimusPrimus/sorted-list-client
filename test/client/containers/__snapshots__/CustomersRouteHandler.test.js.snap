// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersRouteHandler tests Renders 1`] = `
<div
  class="panel panel-default row"
>
  <div
    class="col-xs-0 col-sm-1 col-md-2"
  />
  <section
    class="col-xs-12 col-sm-10 col-md-8"
  >
    <h2
      class="text-primary text-center"
    >
      Customers list
    </h2>
    <div
      class="well row"
    >
      <table
        class="table table-striped table-bordered"
      >
        <thead>
          <tr>
            <th
              class="col-xs-1 text-center actionable order-asc"
            >
              ID
            </th>
            <th
              class="col-xs-3 text-center actionable"
            >
              First Name
            </th>
            <th
              class="col-xs-3 text-center actionable"
            >
              Last Name
            </th>
            <th
              class="col-xs-2 text-center actionable"
            >
              Active
            </th>
            <th
              class="col-xs-3 text-center actionable"
            >
              Joined on
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="text-center"
            >
              00001
            </td>
            <td
              class="text-center"
            >
              MARY
            </td>
            <td
              class="text-center"
            >
              SMITH
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00002
            </td>
            <td
              class="text-center"
            >
              PATRICIA
            </td>
            <td
              class="text-center"
            >
              JOHNSON
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00003
            </td>
            <td
              class="text-center"
            >
              LINDA
            </td>
            <td
              class="text-center"
            >
              WILLIAMS
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00004
            </td>
            <td
              class="text-center"
            >
              BARBARA
            </td>
            <td
              class="text-center"
            >
              JONES
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00005
            </td>
            <td
              class="text-center"
            >
              ELIZABETH
            </td>
            <td
              class="text-center"
            >
              BROWN
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00006
            </td>
            <td
              class="text-center"
            >
              JENNIFER
            </td>
            <td
              class="text-center"
            >
              DAVIS
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00007
            </td>
            <td
              class="text-center"
            >
              MARIA
            </td>
            <td
              class="text-center"
            >
              MILLER
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00008
            </td>
            <td
              class="text-center"
            >
              SUSAN
            </td>
            <td
              class="text-center"
            >
              WILSON
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            >
              2006-02-15
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00009
            </td>
            <td
              class="text-center"
            >
              MARGARET
            </td>
            <td
              class="text-center"
            >
              MOORE
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            />
          </tr>
          <tr>
            <td
              class="text-center"
            >
              00010
            </td>
            <td
              class="text-center"
            >
              DOROTHY
            </td>
            <td
              class="text-center"
            >
              TAYLOR
            </td>
            <td
              class="text-center"
            >
              YES
            </td>
            <td
              class="text-center"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </section>
  <div
    class="col-xs-0 col-sm-1 col-md-2"
  />
</div>
`;
